course/src/matrix3.cpp

```diff
@@ -0,0 +1,208 @@
+                                  (*this)[1][2] * (*this)[2][0]);
+  third_value = (*this)[0][2] *
+                ((*this)[1][0] * (*this)[2][1] - (*this)[1][1] * (*this)[2][0]);
+  return first_value - second_value + third_value;
+}
+
+Matrix3& Matrix3::operator+=(const Matrix3& r_matrix) {
+  row_1_ = row_1_ + r_matrix[0];
+  row_2_ = row_2_ + r_matrix[1];
+  row_3_ = row_3_ + r_matrix[2];
+  return *this;
+}
+
+Matrix3 Matrix3::operator+(const Matrix3& r_matrix) const {
+  Matrix3 aux{*this};
+  aux += r_matrix;
+  return aux;
+}
+
+Matrix3& Matrix3::operator-=(const Matrix3& r_matrix) {
+  row_1_ = row_1_ - r_matrix[0];
+  row_2_ = row_2_ - r_matrix[1];
+  row_3_ = row_3_ - r_matrix[2];
+  return *this;
+}
+
+Matrix3 Matrix3::operator-(const Matrix3& r_matrix) const {
+  Matrix3 aux{*this};
+  aux -= r_matrix;
+  return aux;
+}
+
+Matrix3& Matrix3::operator*=(const double& value) {
+  row_1_ = row_1_ * value;
+  row_2_ = row_2_ * value;
+  row_3_ = row_3_ * value;
+  return *this;
+}
+
+Matrix3 Matrix3::operator*(const double& value) const {
+  Matrix3 aux{*this};
+  aux *= value;
+  return aux;
+}
+
+Matrix3& Matrix3::operator*=(const Matrix3& r_matrix) {
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      (*this)[i][j] = (*this)[i][j] * r_matrix[i][j];
+    }
+  }
+  return *this;
+}
+
+Matrix3 Matrix3::operator*(const Matrix3& r_matrix) const {
+  Matrix3 aux{*this};
+  aux *= r_matrix;
+  return aux;
+}
+
+Vector3 Matrix3::operator*(const Vector3& r_vector) const {
+  Vector3 aux;
+  for (int i = 0; i < 3; ++i) {
+    aux[i] = (*this)[i].dot(r_vector);
+  }
+  return aux;
+}
+
+Matrix3& Matrix3::operator/=(const Matrix3& r_matrix) {
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      if (r_matrix[i][j] == 0) {
+        throw std::invalid_argument("Try to divide by 0");
+      }
+      (*this)[i][j] = (*this)[i][j] / r_matrix[i][j];
+    }
+  }
+  return *this;
+}
+
+Matrix3 Matrix3::operator/(const Matrix3& r_matrix) const {
+  Matrix3 aux{*this};
+  aux /= r_matrix;
+  return aux;
+}
+
+Matrix3& Matrix3::operator/=(const double& value) {
+  if (value == 0) {
+    throw std::invalid_argument("Try to divide by 0");
+  }
+  row_1_ = row_1_ / value;
+  row_2_ = row_2_ / value;
+  row_3_ = row_3_ / value;
+  return *this;
+}
+
+Vector3 Matrix3::row(const int& value) const {
+  Vector3 aux{(*this)[value]};
+  return aux;
+}
+
+Vector3 Matrix3::col(const int& value) const {
+  Vector3 aux;
+  for (int i = 0; i < 3; ++i) {
+    aux[i] = (*this)[i][value];
+  }
+  return aux;
+}
+
+Matrix3 operator*(const double value, const Matrix3& r_matrix) {
+  return r_matrix * value;
+}
+
+std::string Matrix3::row_print(const Vector3& r_vector) const {
+  std::string aux;
+  aux = "[" + std::to_string(static_cast<int>(r_vector.x())) + ", " +
+        std::to_string(static_cast<int>(r_vector.y())) + ", " +
+        std::to_string(static_cast<int>(r_vector.z())) + "]";
+  return aux;
+}
```

Comment:
As we discussed F2F I would expect that this method to change.

Reply:
I was trying to use this as a auxiliary function:

```
std::string row_print(const Vector3& r_vector) {
  std::stringstream ss;
  ss << "[" << r_vector.x() << ", " << r_vector.y() << ", " << r_vector.z() << "]";
  std::string aux = ss.str();  
  return aux;
}
```

And this is the error that is putting when i try to compile:

```
/course/src/matrix3.cpp:15:21: warning: calling a function that uses a default argument is disallowed [fuchsia-default-arguments]
  std::stringstream ss;
                    ^
/usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/sstream:703:26: note: default parameter was declared here
      basic_stringstream(ios_base::openmode __m = ios_base::out | ios_base::in)
                         ^
/course/src/matrix3.cpp:17:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
/course/src/matrix3.cpp:22:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
Done processing /course/src/matrix3.cpp
Total errors found: 2
CMakeFiles/isometry.dir/build.make:110: recipe for target 'CMakeFiles/isometry.dir/src/matrix3.cpp.o' failed
make[2]: *** [CMakeFiles/isometry.dir/src/matrix3.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/isometry.dir/all' failed
make[1]: *** [CMakeFiles/isometry.dir/all] Error 2
Makefile:94: recipe for target 'all' failed
make: *** [all] Error 2
```

Reply:
That is happening because clang-tidy doesn't allow to call a function/constructor that has a default argument.
See the following:
https://en.cppreference.com/w/cpp/io/basic_stringstream/basic_stringstream
See the first constructor, that is the one you are calling.

Try declaring it this way:
```c++
std::stringstream ss(std::ios_base::in | std::ios_base::out );
```

Reply:
Thanks! Working now!